rust_snuba/src/factory.rs

```diff
@@ -0,0 +1,139 @@
+                self.storage_config.clickhouse_table_name.clone(),
+                self.skip_write,
+                2,
+            )),
+            accumulator,
+            BytesInsertBatch { rows: vec![] },
+            self.max_batch_size,
+            self.max_batch_time,
+        );
+
+        match (
+            self.use_rust_processor,
+            processors::get_processing_function(
+                &self.storage_config.message_processor.python_class_name,
+            ),
+        ) {
+            (true, Some(func)) => {
+                struct MessageProcessor {
+                    func: fn(
+                        KafkaPayload,
+                        KafkaMessageMetadata,
+                    ) -> Result<BytesInsertBatch, BadMessage>,
+                }
+
+                impl TaskRunner<KafkaPayload, BytesInsertBatch> for MessageProcessor {
+                    fn get_task(
+                        &self,
+                        message: Message<KafkaPayload>,
+                    ) -> RunTaskFunc<BytesInsertBatch> {
+                        let func = self.func;
+
+                        Box::pin(async move {
+                            let broker_message = match message.inner_message {
+                                InnerMessage::BrokerMessage(msg) => msg,
+                                _ => panic!("Unexpected message type"),
+                            };
+
+                            let metadata = KafkaMessageMetadata {
+                                partition: broker_message.partition.index,
+                                offset: broker_message.offset,
+                                timestamp: broker_message.timestamp,
+                            };
+
+                            match func(broker_message.payload, metadata) {
+                                Ok(transformed) => Ok(Message {
+                                    inner_message: InnerMessage::BrokerMessage(BrokerMessage {
+                                        payload: transformed,
+                                        partition: broker_message.partition,
+                                        offset: broker_message.offset,
+                                        timestamp: broker_message.timestamp,
+                                    }),
+                                }),
+                                Err(_e) => Err(InvalidMessage {
+                                    partition: broker_message.partition,
+                                    offset: broker_message.offset,
+                                }),
+                            }
+                        })
+                    }
+                }
+
+                let task_runner = MessageProcessor { func };
+                Box::new(RunTaskInThreads::new(
+                    next_step,
+                    Box::new(task_runner),
+                    self.concurrency,
+                    Some("process_message"),
+                ))
```

Comment:
Am I reading this wrong or are you dispatching each individual message to a thread without any batching ? 
Wouldn't that imply taking and releasing a lock for every message? Otherwise how are threads going to synchronize? I think there can be a huge performance gain in not having suck lock contention.

Reply:
Haven't noticed any performance issues with this approach yet.

From the tokio docs:
> Tasks are the unit of execution managed by the scheduler. Spawning the task submits it to the Tokio scheduler, which then ensures that the task executes when it has work to do. The spawned task may be executed on the same thread as where it was spawned, or it may execute on a different runtime thread. The task can also be moved between threads after being spawned.